### ISSUE TITLE
Transition to assertion-based testing

### ISSUE THREAD

**Issue by siddMahen** [HeronMkII/lib-common#18]
The current set of examples is still a useful way to test functionality, but we should begin moving towards a more standardized testing framework based on assertions. The examples we write tend not to test edge cases or rare behaviour because they represent typical use.

I'm thinking of something like:

```
lib-common
 └── README.md
 └── makefile
 └── src
     └── ...
 └── build
     └── ...
 └── lib
 └── test
     └── assert.h
     └── assert.c
     └── run_tests.c
     └── can
         └── can_test_1.c
         └── can_test_2.c
         └── ...
     └── uart
         └── uart_test_1.c
         └── uart_test_2.c
         └── ...
     └── queue
         └── ...
     └── stack
         └── ...
```

Here `run_tests.c` would be some program/script that uploads each test to a board, records the outcome and prints the results to the terminal. I'm imaging each individual test would look something like this:

```
...
#include <test/assert.h>

int main() {
    // some setup
   assert(expected_1 == ret_val_1);
   assert(expected_2 == ret_val_2);
   assert(expected_3 == ret_val_3);
   ...
}
```

It would also be useful to have this testing framework display/record other kinds of statistics, such as how much SRAM and program memory is being used, how long things are taking, etc.

In the future, I would like to rollout different versions of `lib-common`, which could be compared against one another. See also https://semver.org/ .

\cc @vogeldylan @brunofalmeida @bonnieweng1T9 @shrey-p

**Comment by brunofalmeida** [comment]
Would `assert.h` and `assert.c` be custom-made assertion code or a standard library? If it's a standard library, how would we see the results since it uses `stdout`?

**Comment by brunofalmeida** [comment]
Would the `main()` function only be in `run_tests.c` if you just want to run that and have it execute all the tests? If you put a `main()` in each file, wouldn't you have to run each individually?

**Comment by siddMahen** [comment]
@brunofalmeida It would have to be custom for exactly the reason you brought up; `stdout` just doesn't work like it regularly does, or at least I don't know how to make it compatible with UART printing.

**Comment by siddMahen** [comment]
@brunofalmeida I think the `run_tests.c` code would run locally, on your machine. It would upload each program to the AVR, listen to the UART output, load the next test, rinse and repeat. It would also measure program size and speed.

**Comment by vogeldylan** [comment]
What's the first example/test case we can develop for this? Might as well start some initial tests.

**Comment by siddMahen** [comment]
@vogeldylan A great first test would be to make sure that `init_uart` and `init_can` and the like correctly set all the register values of the 32m1.

**Comment by siddMahen** [comment]
If many people contributed to these tests, it would also make sure everyone knew how UART and CAN worked at the internal level, which might be good in general. This would require reading the current UART/CAN code, reading the documentation, and making sure everything is kosher.

**Comment by bonnieweng1T9** [comment]
@jackberezny and I can take on this as well!

**Comment by bonnieweng1T9** [comment]
@BrytniR  @JasonRocck  @jackberezny @mercury0524  This is the assert based testing that we discussed today with Sidd and Bruno, and one of the major projects we will be looking into after heartbeat.

**Comment by siddMahen** [comment]
In the future, the harness could also be modified to report program size info, memory usage info, etc

**Comment by brunofalmeida** [comment]
How is it possible for the test harness to read the 32M1's memory contents? The harness can also read UART output right?

**Comment by siddMahen** [comment]
See #28 for on-going work towards this goal.

**Comment by siddMahen** [comment]
Recently merged #28, which addresses almost all of our testing needs. @vogeldylan @alihaydaroglu

The only remaining aspect of testing which isn't incorporated into libtest is EEPROM testing. This can be added at a later date. I wanted to commit libtest to master quickly so we can begin writing tests to prep for qual; also I think it's more of a nice-to-have feature.
